c_common/models/data_speed_up_packet_gatherer/src/data_speed_up_packet_gatherer.c

```diff
@@ -59,130 +152,410 @@ static uint32_t data[ITEMS_PER_DATA_PACKET];
+static uint chip_y = 0xFFFFFFF; // Not a legal chip coordinate
+static bit_field_t missing_seq_nums_store = NULL;
+uint size_of_bitfield = 0;
+static bool alloc_in_sdram = false;
+static uint total_received_seq_nums = 0;
+static uint last_seen_seq_num = 0;
+static uint start_sdram_address = 0;
+
+//! Human readable definitions of the offsets for multicast key elements.
+//! These act as commands sent to the target extra monitor core.
+typedef enum key_offsets {
+    SDRAM_KEY_OFFSET = 0,
+    DATA_KEY_OFFSET = 1,
+    RESTART_KEY_OFFSET = 2
+} key_offsets;
+
+typedef struct data_in_config_t {
+    uint32_t n_chips;
+    struct chip_key_data_t {
+        uint32_t x_coord;
+        uint32_t y_coord;
+        uint32_t base_key;
+    } chip_to_key[];
+} data_in_config_t;
+
+//-----------------------------------------------------------------------------
+// FUNCTIONS
+//-----------------------------------------------------------------------------
+
+//! \brief sends the SDP message built in the my_msg global
+static inline void send_sdp_message(void) {
+    while (!spin1_send_sdp_msg((sdp_msg_t *) &my_msg, SDP_TIMEOUT)) {
+        spin1_delay_us(MESSAGE_DELAY_TIME_WHEN_FAIL);
+    }
+}
 
-void resume_callback() {
-    time = UINT32_MAX;
+//! \brief sends a multicast (with payload) message to the current target chip
+//! \param[in] command: the key offset, which indicates the command being sent
+//! \param[in] payload: the argument to the command
+static inline void send_mc_message(uint command, uint payload) {
+    uint key = data_in_mc_key_map[chip_x][chip_y] + command;
+    while (spin1_send_mc_packet(key, payload, WITH_PAYLOAD) == 0) {
+        spin1_delay_us(MESSAGE_DELAY_TIME_WHEN_FAIL);
+    }
+}
+
+//! \brief sends multicast messages accordingly for an SDP message
+//! \param[in] data: the actual data from the SDP message
+//! \param[in] n_elements: the number of data items in the SDP message
+//! \param[in] send_sdram_address: bool flag for if we should send the sdram
+//                                 address of this set of data
+//! \param[in] sdram_address: the sdram address where this block of data is
+//                            to be written on
+static void process_sdp_message_into_mc_messages(
+        const uint *data, uint n_elements, bool send_sdram_address,
+        uint sdram_address) {
+    // determine size of data to send
+    log_info("Writing %u elements to 0x%08x", n_elements, sdram_address);
+
+    // send mc message with SDRAM location to correct chip
+    if (send_sdram_address) {
+        send_mc_message(SDRAM_KEY_OFFSET, sdram_address);
+    }
+
+    // send mc messages containing rest of sdp data
+    for (uint data_index = 0; data_index < n_elements; data_index++) {
+        send_mc_message(DATA_KEY_OFFSET, data[data_index]);
+    }
+}
+
+//! \brief creates a store for seq nums in a memory store.
+//! \param[in] max_seq_num: the max seq num expected during this stage
```

Comment:
Looking at the code, `size` is misleading because this is actually the maximum sequence number from the received packet. Renamed to `max_seq` (which makes adding 1 sensible below, since the bitfield needs to be sized to receive that maximum value).